File: sdks/python/tests/test_modules.py
"""Runtime dependency inventory tests (ADR-0041).

The wire key is asserted directly, because the relay reads a top-level ``modules`` string map and a
renamed or nested key would be dropped silently by the parser rather than rejected. And because an
event that carries no inventory must keep its exact previous shape, which a shape-agnostic test would
not catch.
"""

import json
import unittest

import condux
from condux.modules import (
    MAX_MODULES,
    MODULES_INTERVAL_SECONDS,
    collect_modules,
    modules_field,
    set_modules,
)

DSN = "http://pub123@relay.test/7"


def recording_transport():
    sent = []

    def transport(url, headers, body):
        sent.append(body)
        return 202, {}

    return transport, (lambda: [json.loads(body) for body in sent])


class ModulesTests(unittest.TestCase):
    def setUp(self):
        condux.clear_modules()

    def test_declared_modules_ride_the_wire_under_the_sentry_key(self):
        transport, events = recording_transport()
        condux.init(DSN, transport=transport, send_modules=False)
        set_modules({"requests": "2.19.0", "flask": "2.0.0"})

        condux.capture_exception(ValueError("boom"))

        self.assertEqual(
            {"flask": "2.0.0", "requests": "2.19.0"}, events()[0]["modules"]
        )

    def test_an_event_carries_no_modules_key_when_none_were_declared(self):
        transport, events = recording_transport()
        condux.init(DSN, transport=transport, send_modules=False)

        condux.capture_exception(ValueError("boom"))

        self.assertNotIn("modules", events()[0])

    def test_the_inventory_rides_the_first_event_then_waits_out_the_interval(self):
        """What makes the feature affordable.

        The server deduplicates a release's inventory to one row per package per day, so repeating the
        map on every event of a busy project spends bytes for nothing. Remove the interval and this
        still passes on the first event but fails on the second.
        """
        transport, events = recording_transport()
        condux.init(DSN, transport=transport, send_modules=False)
        set_modules({"requests": "2.19.0"})

        condux.capture_exception(ValueError("one"))
        condux.capture_exception(ValueError("two"))
        condux.capture_exception(ValueError("three"))

        with_modules = [e for e in events() if "modules" in e]
        self.assertEqual(1, len(with_modules))

    def test_the_inventory_rides_again_once_the_interval_elapses(self):
        """Repeating matters as much as skipping.

        The event carrying the inventory can be dropped by a rate limit or a quota rejection before
        anything parses it, so one attempt per process would lose that day's inventory outright.
        """
        set_modules({"requests": "2.19.0"})
        start = 1_000_000.0

        self.assertIn("modules", modules_field(start))
        self.assertNotIn("modules", modules_field(start + MODULES_INTERVAL_SECONDS - 1))
        self.assertIn("modules", modules_field(start + MODULES_INTERVAL_SECONDS))

    def test_a_fresh_declaration_does_not_wait_out_the_previous_interval(self):
        set_modules({"requests": "2.19.0"})
        start = 1_000_000.0
        self.assertIn("modules", modules_field(start))

        set_modules({"requests": "2.32.0"})

        self.assertEqual({"requests": "2.32.0"}, modules_field(start + 1)["modules"])

    def test_entries_are_sorted_and_capped_so_every_event_carries_the_same_set(self):
        # Zero padded so lexicographic order is also numeric order, making the survivors predictable.
        set_modules({f"pkg-{i:05d}": "1.0.0" for i in range(MAX_MODULES + 50)})

        names = list(modules_field(1.0)["modules"])

        self.assertEqual(MAX_MODULES, len(names))
        self.assertEqual("pkg-00000", names[0])
        self.assertEqual(sorted(names), names)

    def test_blank_entries_are_dropped_rather_than_reported_as_versions(self):
        set_modules({"good": "1.0.0", "blank": "", "": "2.0.0"})

        self.assertEqual({"good": "1.0.0"}, modules_field(1.0)["modules"])

    def test_clearing_removes_the_key_entirely(self):
        set_modules({"requests": "2.19.0"})
        set_modules(None)

        self.assertEqual({}, modules_field(1.0))

    def test_collect_reads_the_installed_distributions(self):
        """Reads the environment, not a requirements file.

        Asserted against this package's own test environment rather than a fixture, because the point
        is that importlib.metadata sees what is actually installed. Every value must be a non-empty
        string: a None version rendered as "running None" would be worse than no row.
        """
        found = collect_modules()

        self.assertGreater(len(found), 0)
        for name, version in found.items():
            self.assertIsInstance(name, str)
            self.assertIsInstance(version, str)
            self.assertTrue(name and version)

    def test_init_collects_so_the_next_event_carries_the_inventory(self):
        """The two halves joined: collection and carrying are each tested alone above."""
        transport, events = recording_transport()

        condux.init(DSN, transport=transport)
        condux.capture_exception(ValueError("boom"))

        self.assertGreater(len(events()[0]["modules"]), 0)

    def test_send_modules_false_leaves_the_inventory_off_entirely(self):
        transport, events = recording_transport()

        condux.init(DSN, transport=transport, send_modules=False)
        condux.capture_exception(ValueError("boom"))

        self.assertNotIn("modules", events()[0])

    def test_init_opting_out_clears_an_inventory_a_previous_init_collected(self):
        """A re-init must not leave the previous run's inventory attached."""
        transport, events = recording_transport()
        condux.init(DSN, transport=transport)

        condux.init(DSN, transport=transport, send_modules=False)
        condux.capture_exception(ValueError("boom"))

        self.assertNotIn("modules", events()[0])


if __name__ == "__main__":
    unittest.main()

File: sdks/ruby/test/test_modules.rb
# frozen_string_literal: true

require "minitest/autorun"
require "json"
require "condux"

# Runtime dependency inventory tests (ADR-0041).
#
# The wire key is asserted directly, because the relay reads a top-level "modules" string map and a
# renamed or nested key would be dropped silently by the parser rather than rejected.
class TestModules < Minitest::Test
  DSN = "http://pub123@relay.test/7"

  def setup
    Condux::Modules.reset
    @sent = []
  end

  def init(send_modules: true)
    transport = lambda do |_url, _headers, body|
      @sent << JSON.parse(body)
      [202, {}]
    end
    Condux.init(dsn: DSN, transport: transport, sleep: ->(*) {}, send_modules: send_modules)
  end

  def test_an_event_carries_the_loaded_gems_under_the_sentry_key
    init
    Condux.capture_message("hi")

    modules = @sent.first["modules"]
    refute_nil modules, "an event should carry the inventory"
    # This suite has required condux and json, so both are activated by definition.
    assert modules.key?("json"), "expected a gem this test has actually loaded"
    modules.each do |name, version|
      assert_kind_of String, name
      assert_kind_of String, version
      refute_empty version
    end
  end

  # What makes the feature affordable: the server deduplicates a release's inventory to one row per
  # package per day, so repeating the map on every event spends bytes for nothing.
  def test_the_inventory_rides_the_first_event_then_waits_out_the_interval
    init
    3.times { |i| Condux.capture_message("m#{i}") }

    assert_equal 1, @sent.count { |e| e.key?("modules") }
  end

  # Repeating matters as much as skipping: the event carrying the inventory can be dropped by a rate
  # limit before anything parses it, so one attempt per process would lose that day's inventory.
  def test_the_inventory_rides_again_once_the_interval_elapses
    start = 1_000_000.0

    refute_empty Condux::Modules.fields(start)
    assert_empty Condux::Modules.fields(start + Condux::Modules::MODULES_INTERVAL_SECONDS - 1)
    refute_empty Condux::Modules.fields(start + Condux::Modules::MODULES_INTERVAL_SECONDS)
  end

  def test_send_modules_false_leaves_the_inventory_off_entirely
    init(send_modules: false)
    Condux.capture_message("hi")

    refute @sent.first.key?("modules")
  end

  # Collection is lazy for a reason, so the reason is asserted rather than only commented.
  #
  # Gem.loaded_specs reports ACTIVATED gems, not installed ones: in a bare interpreter it holds a
  # single entry, because nothing has been required yet. Collecting during init, which an application
  # calls early in boot, would therefore report almost nothing. What makes capture-time collection
  # correct is that collect reads the LIVE set every time rather than a snapshot taken at init.
  #
  # Asserted by adding a spec to Gem.loaded_specs directly rather than by requiring some library,
  # because which standard libraries are gems differs by Ruby version: on the 2.6 this was written
  # against, requiring "base64" activates nothing at all, so that test would have passed or failed for
  # reasons unrelated to the SDK.
  def test_collection_reads_the_live_set_rather_than_a_snapshot
    init
    refute Condux::Modules.collect.key?("condux-fake-gem")

    spec = Gem::Specification.new do |s|
      s.name = "condux-fake-gem"
      s.version = "9.9.9"
    end
    Gem.loaded_specs["condux-fake-gem"] = spec
    begin
      assert_equal "9.9.9", Condux::Modules.collect["condux-fake-gem"]
    ensure
      Gem.loaded_specs.delete("condux-fake-gem")
    end

    refute Condux::Modules.collect.key?("condux-fake-gem")
  end

  def test_entries_are_sorted_so_every_event_carries_the_same_order
    init
    Condux.capture_message("hi")

    names = @sent.first["modules"].keys
    assert_equal names.sort, names
  end
end
